util/stmtsummary/statement_summary.go

```diff
@@ -388,20 +410,63 @@ func (ssMap *stmtSummaryByDigestMap) SetRefreshInterval(value string, inSession
 	}
 }
 
-func (ssMap *stmtSummaryByDigestMap) RefreshInterval() int64 {
+// refreshInterval gets the refresh interval for summaries.
+func (ssMap *stmtSummaryByDigestMap) refreshInterval() int64 {
 	return atomic.LoadInt64(&ssMap.sysVars.refreshInterval)
 }
 
+// SetHistorySize sets the history size for all summaries.
+func (ssMap *stmtSummaryByDigestMap) SetHistorySize(value string, inSession bool) {
+	ssMap.sysVars.Lock()
+	if inSession {
+		ssMap.sysVars.sessionHistorySize = value
+	} else {
+		ssMap.sysVars.globalHistorySize = value
+	}
+	sessionHistorySize := ssMap.sysVars.sessionHistorySize
+	globalHistorySize := ssMap.sysVars.globalHistorySize
+	ssMap.sysVars.Unlock()
+
+	// Calculate the cached `historySize`.
+	var size int
+	var err error
+	if ssMap.isSet(sessionHistorySize) {
+		size, err = strconv.Atoi(sessionHistorySize)
+		if err != nil {
+			size = 0
+		}
+	}
+	if size <= 0 {
+		size, err = strconv.Atoi(globalHistorySize)
+		if err != nil {
+			size = 0
+		}
+	}
+	if size > 0 {
+		atomic.StoreInt32(&ssMap.sysVars.historySize, int32(size))
+	}
+}
+
+// historySize gets the history size for summaries.
+func (ssMap *stmtSummaryByDigestMap) historySize() int {
+	size := int(atomic.LoadInt32(&ssMap.sysVars.historySize))
+	if size <= 0 {
```

Comment:
Redundant check?
Already do the check when setting the value. line#445

Reply:
Agree, we should set history size in `SetHistorySize` at once.

Reply:
It's normal that `SetHistorySize` is always called after bootstrap. But I do the check just in case that there would be any bug that prevents `SetHistorySize` be called after bootstrap, and it'll be a big problem.

Reply:
Now I find out the problem. 
If TiDB is just upgraded, global variables stored in TiKV doesn't include `tidb_stmt_summary_history_size`, so the loaded `GlobalVariableCache` doesn't include this variable, and then `historySize` won't be initialized.
I'll fix it.

Reply:
Addressed.